### PR TITLE
[grok2] Fix silently incomplete expert weights when tp_size < presharded checkpoint width

### PR DESCRIPTION
**Pull request by sushildubey171**
## Motivation

`xai-org/grok-2` ships a checkpoint that is **presharded 8 ways**: the MoE expert weights are split across `pytorch_model-*-TP-000..007.safetensors`, while every other tensor lives in `*-TP-common.safetensors` as a full tensor.

`_prepare_presharded_weights` has each rank glob only its own shard:

```python
allow_patterns += [f"*-TP-{tp_rank:03d}.safetensors", "*-TP-common.safetensors"]
```

At `tp_size < 8` this reads only shard `tp_rank` of each expert group while the layers are sized for the smaller TP. **It does not raise**: `hit_names` counts tensor *names*, and every name is present in the rank's single shard, so the load reports `#missing_exclude_scales: 0` and the server starts normally on a partially-populated model. The failure is silent and shows up only as degraded quality.

Measured on 8x MI350X (gfx950), `tp_size=4`, FP8, GSM8K 5-shot / 1319 questions:

| | GSM8K |
|---|---|
| `tp_size=4`, current `main` | **0.901** |
| `tp_size=4`, this PR | **0.926** |
| `tp_size=8` (reference, unaffected by this PR) | 0.932 |

The model code itself is already TP-agnostic (attention only asserts divisibility) - the limitation is purely checkpoint file selection.

## Modification

Rank `r` at `tp_size=T` owns exactly the contiguous slice covered by checkpoint shards `[r*g, (r+1)*g)` where `g = N/T`, so its weights can be reconstructed by concatenating those `g` shards.

1. **File selection** - derive the checkpoint width from `max(shard id) + 1` and reject gaps, then have each rank select its `g` shards. `tp_size > N` would require *splitting* a shard and is rejected with an explicit error.
2. **Merge-aware weights iterator** - selecting extra files is **not sufficient on its own**: the shards expose *identical tensor names*, so the default iterator would yield duplicates and keep only one. The new iterator opens a group's shards together (via `ExitStack`) and concatenates per tensor, keeping peak memory at one merged tensor. The concat axis is taken from the tensor name (`w1`/`w3` column-parallel -> dim 0, `w2` row-parallel -> dim 1) and anything else raises rather than guessing, since a wrong axis would yield a plausibly shaped but corrupted tensor.

Both helpers fall back to the stock loader when the checkpoint is not presharded, and are installed for every `tp_size` (a presharded checkpoint has to be reassembled at `tp_size=1` too).

## Correctness validation

- **Bit-exact weight reconstruction** - the per-rank tensors produced by this patch are `torch.equal` to an independently merged checkpoint sliced for `tp_size=4`, verified across `w1`/`w2`/`w3` and all 4 ranks.
- **No-op at native TP** - with the patch applied, `tp_size=8` output is **token-identical** to stock on a fixed greedy prompt set, and `#hit_names` is unchanged (707).
- **Memory** - `tp_size=4` uses 66.10 GB/GPU vs 35.40 GB/GPU at `tp_size=8`, the expected 2x.

## Checklist

- [x] Format with `black` / `isort`
- [x] Behaviour unchanged when `tp_size == N`
- [x] Clear error for unsupported `tp_size` and for incomplete checkpoints
- [x] Task-level accuracy verified (GSM8K, 1319 questions)

---
### CI States

Latest PR Test (Base): :x: [Run #33039516363](https://github.com/sgl-project/sglang/actions/runs/33039516363)
Latest PR Test (Extra): :x: [Run #33039516174](https://github.com/sgl-project/sglang/actions/runs/33039516174)
Latest PR Test (AMD ROCm 7.2): :x: [Run #33039516338](https://github.com/sgl-project/sglang/actions/runs/33039516338)

### CODE REVIEW

**Comment by sushildubey171** [comment]
@merrymercy  @BBuf  @Fridge003 please review, thanks!

File: python/sglang/srt/models/grok.py [hunk 1]
```diff
@@ -990,6 +1033,104 @@ def _prepare_presharded_weights(
...
+    for prefix, items in groups.items():
+        items.sort()  # ascending checkpoint-shard index => correct concat order
+        if len(items) > 1:
+            merged[prefix] = [p for _, p in items]
+        else:
+            single.append(items[0][1])
+    return merged, passthrough + single
+
+
+def _shard_concat_dim(name, shape, hidden_size):
```
Comment (hubertlu-tw) [inline python/sglang/srt/models/grok.py]: The fallback axis inference can silently reconstruct an unexpected tensor on the wrong dimension. The checkpoint currently contains only `w1`, `w2`, and  `w3`, whose axes are known. Please fail clearly for any other tensor rather than producing a potentially valid-shaped but corrupted model. This also removes the now-unnecessary `hidden_size` plumbing.

File: python/sglang/srt/models/grok.py [hunk 2]
```diff
@@ -967,13 +972,51 @@ def _prepare_presharded_weights(
...
+    tp_size = get_parallel().tp_size
+
+    # How many ways is this checkpoint presharded? (e.g. 8 for xai-org/grok-2)
+    n_ckpt_shards = len(
+        {
+            m.group(1)
+            for f in glob.glob(os.path.join(hf_folder, "*-TP-*.safetensors"))
+            if (m := re.search(r"-TP-(\d{3})\.safetensors$", os.path.basename(f)))
+        }
+    )
```
Comment (hubertlu-tw) [inline python/sglang/srt/models/grok.py]: `len(set(shard_ids))` gives the number of distinct IDs, not the checkpoint width. For example, IDs `{000, 002}` are counted as a 2-way checkpoint and lead the loader to request nonexistent shard `001`. Please derive the width from the maximum ID and reject gaps before selecting rank files.

File: python/sglang/srt/models/grok.py [hunk 3]
```diff
@@ -723,6 +723,11 @@ def __init__(
 
         if get_parallel().tp_size > 1:
             setattr(DefaultModelLoader, "_prepare_weights", _prepare_presharded_weights)
+            setattr(
+                DefaultModelLoader,
+                "_get_weights_iterator",
+                _presharded_get_weights_iterator,
+            )
```
Comment (hubertlu-tw) [inline python/sglang/srt/models/grok.py]: The new error tells users to choose any `tp_size` that divides the checkpoint width, but TP=1 still takes the old path. That path sees duplicate tensor names and does not reconstruct the eight shards. Either explicitly reject TP=1 and narrow the error message, or install the merge-aware path for TP=1 too. The following keeps the general divisor behavior.

File: python/sglang/srt/models/grok.py [hunk 4]
```diff
@@ -990,6 +1033,104 @@ def _prepare_presharded_weights(
...
+                        f"presharded files for {prefix} expose different tensor names"
+                    )
+            for name in keys:
+                parts = [h.get_tensor(name) for h in handles]
+                dim = _shard_concat_dim(name, list(parts[0].shape), hidden_size)
+                yield name, torch.cat(parts, dim=dim)
+                del parts
+        finally:
+            for h in handles:
+                h.__exit__(None, None, None)
```
Comment (hubertlu-tw) [inline python/sglang/srt/models/grok.py]: Please use the `safe_open` context-manager protocol instead of manually calling `__exit__` on handles that were never entered. `ExitStack` preserves the current requirement to keep every shard open while each tensor is concatenated, while guaranteeing cleanup if opening or reading a later shard fails.

**Review by hubertlu-tw — CHANGES_REQUESTED** [review]

**Comment by hubertlu-tw** [comment]
Thanks for the fix. I have left some comments on your code changes.

Additionally, I have also tested your code changes with the following commands on my MI355 systems:
```
HIP_VISIBLE_DEVICES=0,1,2,3 SGLANG_INT4_WEIGHT=0 sglang serve \
  --model-path xai-org/grok-2 \
  --host 127.0.0.1 \
  --port 30000 \
  --tp 4 \
  --quantization fp8 \
  --attention-backend aiter \
  --mem-fraction-static 0.85 \
  --tokenizer-path alvarobartt/grok-2-tokenizer \
  --trust-remote-code
```

```
cd /sgl-workspace/sglang && python3 benchmark/gsm8k/bench_sglang.py \
  --num-questions 1319 \
  --parallel 1319 \
  --num-shots 5
```

(pre-PR) Accuracy: 0.909 / 0.906 --> (post-PR) Accuracy: 0.923 / 0.926

cc: @HaiShaw

File: python/sglang/srt/models/grok.py (hunk 1 above)
Comment (sushildubey171) [reply python/sglang/srt/models/grok.py]: Agreed — this was exactly the silent-corruption class I flagged in the description, so thanks for catching it in my own code. _shard_concat_dim now raises for anything other than w1/w2/w3, and the hidden_size plumbing is gone.

File: python/sglang/srt/models/grok.py (hunk 2 above)
Comment (sushildubey171) [reply python/sglang/srt/models/grok.py]: Good catch — fixed. Width is now `max(shard id) + 1` with an explicit gap check that names the missing shards, so `{000, 002}` reports an incomplete checkpoint instead of being read as 2-way.

File: python/sglang/srt/models/grok.py (hunk 3 above)
Comment (sushildubey171) [reply python/sglang/srt/models/grok.py]: You're right, and I went with the general-divisor option: the early `tp_size == 1` return is gone and both helpers are installed for every tp_size, so a presharded checkpoint is reassembled at `TP=1` too. They still fall back to the stock loader when the checkpoint isn't presharded, so non-Grok paths are untouched.

File: python/sglang/srt/models/grok.py (hunk 4 above)
Comment (sushildubey171) [reply python/sglang/srt/models/grok.py]: Fixed — shards are opened through contextlib.ExitStack. It also closes the leak I had if safe_open failed partway through the list.

**Comment by sushildubey171** [comment]
> Thanks for the fix. I have left some comments on your code changes.
> 
> Additionally, I have also tested your code changes with the following commands on my MI355 systems:
> 
> ```
> HIP_VISIBLE_DEVICES=0,1,2,3 SGLANG_INT4_WEIGHT=0 sglang serve \
>   --model-path xai-org/grok-2 \
>   --host 127.0.0.1 \
>   --port 30000 \
>   --tp 4 \
>   --quantization fp8 \
>   --attention-backend aiter \
>   --mem-fraction-static 0.85 \
>   --tokenizer-path alvarobartt/grok-2-tokenizer \
>   --trust-remote-code
> ```
> 
> ```
> cd /sgl-workspace/sglang && python3 benchmark/gsm8k/bench_sglang.py \
>   --num-questions 1319 \
>   --parallel 1319 \
>   --num-shots 5
> ```
> 
> (pre-PR) Accuracy: 0.909 / 0.906 --> (post-PR) Accuracy: 0.923 / 0.926
> 
> cc: @HaiShaw

Thanks @hubertlu-tw for the review and for independently reproducing this on MI355.
